File: examples/basic_usage.cpp
#include <algoat/algoat.hpp>
#include <iostream>
#include <vector>
#include <fstream>
#include <span>

int main() {
    std::cout << "--- Algoat Basic Usage Example ---\n\n";

    // 1. Without Config (Auto Dispatch)
    std::cout << "[Auto Dispatch]\n";
    std::vector<int> data = {42, 7, 19, 100, 3, 55, 21};
    
    std::cout << "Original data: ";
    for (int v : data) std::cout << v << " ";
    std::cout << "\n";

    algoat::sort(std::span{data});
    
    std::cout << "Sorted data:   ";
    for (int v : data) std::cout << v << " ";
    std::cout << "\n";

    int target = 19;
    auto res = algoat::search(std::span{data}, target);
    if (res) {
        std::cout << "Found " << target << " at index " << *res << "\n";
    } else {
        std::cout << "Could not find " << target << "\n";
    }

    // 2. With Config Override
    std::cout << "\n[Config Override]\n";
    
    std::ofstream config_file("example_config.json");
    config_file << R"({
        "sorting": {"prefer": "heapsort"},
        "searching": {"prefer": "linearsearch"}
    })";
    config_file.close();

    algoat::load_global_config("example_config.json");

    std::vector<int> data2 = {9, 1, 8, 2, 7, 3, 6, 4, 5};
    algoat::sort(std::span{data2});
    
    std::cout << "Sorted data (heapsort): ";
    for (int v : data2) std::cout << v << " ";
    std::cout << "\n";

    int target2 = 7;
    auto res2 = algoat::search(std::span{data2}, target2);
    if (res2) {
        std::cout << "Found " << target2 << " at index " << *res2 << " (linearsearch)\n";
    }

    return 0;
}

File: examples/basic_usage.py
#!/usr/bin/env python3
"""Algoat Basic Usage Example — Python Edition

Demonstrates sorting, searching, and config-based algorithm overrides
using the algoat Python bindings (backed by C++20 via nanobind).
"""

import json
import tempfile
import os
import algoat


def main():
    print("--- Algoat Basic Usage Example (Python) ---\n")

    # 1. Auto Dispatch (no config)
    print("[Auto Dispatch]")
    data = [42, 7, 19, 100, 3, 55, 21]
    print(f"Original data: {data}")

    sorted_data = algoat.sort(data)
    print(f"Sorted data:   {sorted_data}")

    target = 19
    index = algoat.search(sorted_data, target)
    if index is not None:
        print(f"Found {target} at index {index}")
    else:
        print(f"Could not find {target}")

    # 2. Config Override
    print("\n[Config Override]")
    config = {
        "sorting": {"prefer": "heapsort"},
        "searching": {"prefer": "linearsearch"},
    }

    config_path = os.path.join(tempfile.gettempdir(), "algoat_example_config.json")
    with open(config_path, "w") as f:
        json.dump(config, f)

    algoat.load_global_config(config_path)

    data2 = [9, 1, 8, 2, 7, 3, 6, 4, 5]
    sorted_data2 = algoat.sort(data2)
    print(f"Sorted data (heapsort): {sorted_data2}")

    target2 = 7
    index2 = algoat.search(sorted_data2, target2)
    if index2 is not None:
        print(f"Found {target2} at index {index2} (linearsearch)")

    # 3. Float support
    print("\n[Float Support]")
    floats = [3.14, 1.41, 2.72, 0.58]
    sorted_floats = algoat.sort(floats)
    print(f"Sorted floats: {sorted_floats}")

    # Cleanup
    os.remove(config_path)
    print("\nDone!")


if __name__ == "__main__":
    main()
